src/vocab/mod.rs

```diff
@@ -157,3 +160,151 @@ pub(crate) fn bracket(word: &str) -> String {
 
     bracketed
 }
+
+/// Cutoff to determine vocabulary size.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize)]
+#[serde(tag = "type", content = "value")]
+pub enum Cutoff {
+    /// Cutoff based on minimum frequency, items appearing less than
+    /// `min_count` times are discarded.
+    MinCount(usize),
+    /// Cutoff based on a target size, up to `target_size` items are kept
+    /// in the vocabulary. If the item at `target_size+1` appears `n` times,
+    /// all items with frequency `n` and smaller are discarded.
+    TargetSize(usize),
+}
+
+impl Cutoff {
+    pub(crate) fn filter<T, S>(
+        &self,
+        items: impl IntoIterator<Item = (T, usize)>,
+    ) -> Vec<CountedType<S>>
+    where
+        T: Hash + Eq + Into<S>,
+        S: Hash + Eq + Clone + Ord,
+    {
+        match self {
+            Cutoff::MinCount(min_count) => filter_minfreq(items, *min_count),
+            Cutoff::TargetSize(target_size) => filter_targetsize(items, *target_size),
+        }
+    }
+}
+
+fn filter_minfreq<T, S>(
+    items: impl IntoIterator<Item = (T, usize)>,
+    min_count: usize,
+) -> Vec<CountedType<S>>
+where
+    T: Hash + Eq + Into<S>,
+    S: Hash + Eq + Clone + Ord,
+{
+    let mut types: Vec<_> = items
+        .into_iter()
+        .filter(|(_, count)| *count >= min_count as usize)
+        .map(|(item, count)| CountedType::new(item.into(), count))
+        .collect();
+    types.sort_unstable_by(|w1, w2| w2.cmp(&w1));
+    types
+}
+
+fn filter_targetsize<T, S>(
+    items: impl IntoIterator<Item = (T, usize)>,
+    target_size: usize,
+) -> Vec<CountedType<S>>
+where
+    T: Hash + Eq + Into<S>,
+    S: Hash + Eq + Clone + Ord,
+{
+    let mut items = items
+        .into_iter()
+        .map(|(item, count)| CountedType::new(item.into(), count))
+        .collect::<Vec<_>>();
+    items.sort_unstable_by(|i1, i2| i2.cmp(&i1));
+    if target_size > items.len() {
+        return items;
+    }
+    // *_bound_* methods expect lists sorted by items' order, we are sorting in reverse so
+    // we need to simulate the natural ordering.
+    let cutoff_idx =
+        items.lower_bound_by(|i| (-(i.count as isize)).cmp(&-(items[target_size].count as isize)));
```

Comment:
Hard to read, but I think something like

```suggestion
        items.lower_bound_by_key(Reverse(i.count), |key| Reverse(items[target_size].count));
```

should work.